src/lib/hmacsha1.c

```diff
@@ -28,119 +26,15 @@ unsigned int sha1_data_problems = 0;
 void fr_hmac_sha1(uint8_t digest[SHA1_DIGEST_LENGTH], uint8_t const *text, size_t text_len,
 		  uint8_t const *key, size_t key_len)
 {
-	fr_sha1_ctx context;
-	uint8_t k_ipad[65];    /* inner padding - key XORd with ipad */
-	uint8_t k_opad[65];    /* outer padding - key XORd with opad */
-	uint8_t tk[20];
-	int i;
-	/* if key is longer than 64 bytes reset it to key=SHA1(key) */
-	if (key_len > 64) {
-
-		fr_sha1_ctx      tctx;
-
-		fr_sha1_init(&tctx);
-		fr_sha1_update(&tctx, key, key_len);
-		fr_sha1_final(tk, &tctx);
-
-		key = tk;
-		key_len = 20;
-	}
-
-#ifdef HMAC_SHA1_DATA_PROBLEMS
-	if(sha1_data_problems)
-	{
-		int j,k;
-
-		printf("\nhmac-sha1 key(%d): ", key_len);
-		j=0; k=0;
-		for (i = 0; i < key_len; i++) {
-			if(j==4) {
-				printf("_");
-				j=0;
-			}
-			j++;
-
-			printf("%02x", key[i]);
-		}
-		printf("\nDATA: (%d)    ",text_len);
-
-		j=0; k=0;
-		for (i = 0; i < text_len; i++) {
-		  if(k==20) {
-		    printf("\n	    ");
-		    k=0;
-		    j=0;
-		  }
-		  if(j==4) {
-		    printf("_");
-		    j=0;
-		  }
-		  k++;
-		  j++;
-
-		  printf("%02x", text[i]);
-		}
-		printf("\n");
-	}
-#endif
+	HMAC_CTX *ctx  = HMAC_CTX_new();
 
-
-	/*
-	 * the HMAC_SHA1 transform looks like:
-	 *
-	 * SHA1(K XOR opad, SHA1(K XOR ipad, text))
-	 *
-	 * where K is an n byte key
-	 * ipad is the byte 0x36 repeated 64 times
-
-	 * opad is the byte 0x5c repeated 64 times
-	 * and text is the data being protected
-	 */
-
-	/* start out by storing key in pads */
-	memset(k_ipad, 0, sizeof(k_ipad));
-	memset(k_opad, 0, sizeof(k_opad));
-	memcpy(k_ipad, key, key_len);
-	memcpy(k_opad, key, key_len);
-
-	/* XOR key with ipad and opad values */
-	for (i = 0; i < 64; i++) {
-		k_ipad[i] ^= 0x36;
-		k_opad[i] ^= 0x5c;
-	}
-	/*
-	 * perform inner SHA1
-	 */
-	fr_sha1_init(&context);				/* init context for 1st pass */
-	fr_sha1_update(&context, k_ipad, 64);		/* start with inner pad */
-	fr_sha1_update(&context, text, text_len);	/* then text of datagram */
-	fr_sha1_final(digest, &context);		/* finish up 1st pass */
-	/*
-	 * perform outer SHA1
-	 */
-	fr_sha1_init(&context);				/* init context for 2nd pass */
-	fr_sha1_update(&context, k_opad, 64);		/* start with outer pad */
-	fr_sha1_update(&context, digest, 20);		/* then results of 1st hash */
-	fr_sha1_final(digest, &context);		/* finish up 2nd pass */
-
-#ifdef HMAC_SHA1_DATA_PROBLEMS
-	if (sha1_data_problems) {
-		int j;
-
-		printf("\nhmac-sha1 mac(20): ");
-		j=0;
-		for (i = 0; i < 20; i++) {
-			if(j==4) {
-				printf("_");
-				j=0;
-			}
-			j++;
-
-			printf("%02x", digest[i]);
-		}
-		printf("\n");
-	}
+#ifdef EVP_MD_CTX_FLAG_NON_FIPS_ALLOW
```

Comment:
I believe SHA-1 to be allowed by FIPS 140-2 so perhaps I can drop these lines (I mirrored the MD5 code closely), but I'll take an AI to investigate this.